cpp/src/arrow/filesystem/s3fs_benchmark.cc

```diff
@@ -0,0 +1,317 @@
+      ASSERT_OK_AND_ASSIGN(buf, file->ReadAt(offset, read));
+      total_bytes += buf->size();
+      offset += buf->size();
+    }
+  }
+  st.SetBytesProcessed(total_bytes);
+  st.SetItemsProcessed(total_items);
+  std::cout << "Read the file " << total_items << " times" << std::endl;
+}
+
+/// Read the file in small chunks, but using read coalescing.
+static void CoalescedRead(benchmark::State& st, S3FileSystem* fs,
+                          const std::string& path) {
+  int64_t total_bytes = 0;
+  int total_items = 0;
+  for (auto _ : st) {
+    std::shared_ptr<io::RandomAccessFile> file;
+    std::shared_ptr<Buffer> buf;
+    int64_t size = 0;
+    ASSERT_OK_AND_ASSIGN(file, fs->OpenInputFile(path));
+    ASSERT_OK_AND_ASSIGN(size, file->GetSize());
+    total_items += 1;
+
+    io::internal::ReadRangeCache cache(file, 8192, 64 * 1024 * 1024);
+    std::vector<io::ReadRange> ranges;
+
+    int64_t offset = 0;
+    while (offset < size) {
+      const int64_t read = std::min(size, kChunkSize);
+      ranges.push_back(io::ReadRange{offset, read});
+      offset += read;
+    }
+    ASSERT_OK(cache.Cache(ranges));
+
+    offset = 0;
+    while (offset < size) {
+      const int64_t read = std::min(size, kChunkSize);
+      ASSERT_OK_AND_ASSIGN(buf, cache.Read({offset, read}));
+      total_bytes += buf->size();
+      offset += read;
+    }
+  }
+  st.SetBytesProcessed(total_bytes);
+  st.SetItemsProcessed(total_items);
+  std::cout << "Read the file " << total_items << " times" << std::endl;
+}
+
+/// Read a Parquet file from S3.
+static void ParquetRead(benchmark::State& st, S3FileSystem* fs, const std::string& path) {
+  int64_t total_bytes = 0;
+  int total_items = 0;
+  for (auto _ : st) {
+    std::shared_ptr<io::RandomAccessFile> file;
+    int64_t size = 0;
+    // N.B. this is an extra call to S3 and influences the
+    // benchmark. We should compute or cache the size.
+    ASSERT_OK_AND_ASSIGN(file, fs->OpenInputFile(path));
+    ASSERT_OK_AND_ASSIGN(size, file->GetSize());
+
+    parquet::ArrowReaderProperties properties;
+    properties.set_use_threads(true);
+    std::unique_ptr<parquet::arrow::FileReader> reader;
+    parquet::arrow::FileReaderBuilder builder;
+    ASSERT_OK(builder.Open(file));
+    ASSERT_OK(builder.properties(properties)->Build(&reader));
+    std::shared_ptr<RecordBatchReader> rb_reader;
+    ASSERT_OK(reader->GetRecordBatchReader({0}, &rb_reader));
+    std::shared_ptr<Table> table;
+    ASSERT_OK(rb_reader->ReadAll(&table));
+
+    // TODO: actually measure table memory usage
+    total_bytes += size;
+    total_items += 1;
+  }
+  st.SetBytesProcessed(total_bytes);
+  st.SetItemsProcessed(total_items);
+  std::cout << "Read the file " << total_items << " times" << std::endl;
+}
+
+BENCHMARK_DEFINE_F(MinioFixture, ReadAll1Mib)(benchmark::State& st) {
+  NaiveRead(st, fs_.get(), "bucket/bytes_1mib");
+}
+BENCHMARK_REGISTER_F(MinioFixture, ReadAll1Mib)->MinTime(15)->UseRealTime();
```

Comment:
Are 15 seconds necessary for benchmark stability? Usually we don't specify min time in benchmark source and instead let the user pass `--benchmark_min_time` on the command line.

Reply:
In general yes, we need a much higher min time in order to run more than 1 iteration of these benchmarks, but I've removed it in favor of the CLI switch.